src/sentry/static/sentry/app/components/confirm.tsx

```diff
@@ -14,268 +14,259 @@ type MessageRenderProps = {
+    };
 
-    // Close modal
-    this.setState({
-      isModalOpen: false,
-      disableConfirmButton: true,
-    });
-    this.confirming = true;
+    openModal(props => <ConfirmModal {...props} {...modalProps} />);
   };
 
-  handleToggle = (e: React.MouseEvent): void => {
-    const {disabled, bypass, stopPropagation, onConfirm} = this.props;
-    if (disabled) {
-      return;
-    }
+  if (typeof children === 'function') {
+    return children({open: triggerModal});
+  }
 
-    if (e && stopPropagation) {
-      e.stopPropagation();
-    }
+  if (!React.isValidElement(children)) {
+    return null;
+  }
 
-    if (bypass) {
-      onConfirm();
-      return;
-    }
+  // TODO(ts): Understand why the return type of `cloneElement` is strange
+  return React.cloneElement(children, {disabled, onClick: triggerModal}) as any;
+}
 
-    // Current state is closed, means it will toggle open
-    if (!this.state.isModalOpen) {
-      this.openModal();
-    } else {
-      this.closeModal();
-    }
+type ModalProps = ModalRenderProps &
+  Pick<
+    Props,
+    | 'priority'
+    | 'renderMessage'
+    | 'message'
+    | 'confirmText'
+    | 'cancelText'
+    | 'header'
+    | 'onConfirm'
+    | 'onCancel'
+    | 'disableConfirmButton'
+  >;
+
+type ModalState = {
+  /**
+   * Is confirm button disabled
+   */
+  disableConfirmButton: boolean;
+  /**
+   * The callback registered from the rendered message to call
+   */
+  confirmCallback: null | (() => void);
+};
+
+class ConfirmModal extends React.Component<ModalProps, ModalState> {
+  state: ModalState = {
+    disableConfirmButton: !!this.props.disableConfirmButton,
+    confirmCallback: null,
   };
 
+  confirming: boolean = false;
+
   render() {
     const {
-      disabled,
-      message,
-      renderMessage,
+      Header,
+      Body,
+      Footer,
       priority,
+      renderMessage,
+      message,
       confirmText,
       cancelText,
-      children,
       header,
+      onConfirm,
+      onCancel,
+      disableConfirmButton,
+      closeModal,
     } = this.props;
 
-    let confirmMessage: React.ReactNode;
-
-    if (typeof renderMessage === 'function') {
-      confirmMessage = renderMessage({
-        confirm: this.handleConfirm,
-        close: this.handleToggle,
-      });
-    } else {
-      confirmMessage = React.isValidElement(message) ? (
+    const handleClose = () => {
+      onCancel?.();
+      this.setState({disableConfirmButton: disableConfirmButton ?? false});
+
+      // always reset `confirming` when modal visibility changes
+      this.confirming = false;
+      closeModal();
+    };
+
+    const handleConfirm = () => {
+      // `confirming` is used to make sure `onConfirm` is only called once
+      if (!this.confirming) {
+        onConfirm?.();
+        this.state.confirmCallback?.();
+      }
+
+      this.setState({disableConfirmButton: true});
+      this.confirming = true;
+      closeModal();
+    };
+
+    const confirmMessage =
```

Comment:
can we create a function for this? I believe it would be easier to read

Reply:
yeah, good idea